File: CodingKatas/src/main/java/katas/stack/Stack.java
package katas.stack;

import java.util.ArrayList;
import java.util.List;

public class Stack<T> {

    List<T> elements = new ArrayList<>();

    public boolean isEmpty() {
        return elements.isEmpty();
    }

    public void push(T element) {
        this.elements.add(element);
    }

    public T top() {
        if(isEmpty()) throw new IllegalStateException("Stack is empty");
        return elements.get(elements.size() - 1);
    }

    public Stack<T> pop() {
        if(isEmpty()) throw new IllegalStateException("Stack is empty");
        elements.remove(elements.size() - 1);
        return this;
    }
}

File: CodingKatas/src/test/java/katas/stack/stack.pl
/*
            isempty(newstack) = true
            isempty(push(x, l)) = false
            top(push(x, l)) = x
            pop(push(x, l)) = l
            push(top(s),pop(s)) = s
            pop(emptyStack()) = ERROR
            top(emptyStack()) = ERROR
*/

isempty(newstack).
top(stack(X, _), X).
pop(stack(_, L), L).
push(X,S, stack(X,S)).

test:-
    isempty(newstack),
    not(isempty(stack(x, _))),
    top(stack(x, _),x),
    pop(stack(_, L), L),
    push(X,S, stack(X,S)),
    not(pop(newstack, _)),
    not(top(newstack, _)).